File: sdn/ipc-bench-master/unix_lat_client.c
#include <sys/socket.h>
#include <sys/un.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

char *socket_path = "/var/sdn/bench.socket";
// char *socket_path = "\0hidden";
struct sockaddr_un addr;


int main(int argc, char *argv[]) {
  int fd; /* the pair of socket descriptors */
  int expect_recv_size = 0;
  int expect_send_size = 0;
  char buf[1024];
  int64_t count, i, delta;
#ifdef HAS_CLOCK_GETTIME_MONOTONIC
  struct timespec start, stop;
#else
  struct timeval start, stop;
#endif

  if (argc != 4) {
    printf("usage: ./unix_lat_client \"command\" <expect_recv_size> <roundtrip-count>\n");
    // printf("usage: unix_lat 22 <roundtrip-count>\n");
    return 1;
  }

  expect_send_size = strlen(argv[1])+1;
  expect_recv_size = atoi(argv[2]);
  count = atol(argv[3]);

  // buf = malloc(size);
  // memset(&buf, 'o', sizeof(buf)); buf[size-1]=0;
  if (buf == NULL) {
    perror("malloc");
    return 1;
  }

  printf("message content: %s \n", argv[1]);
  printf("message size: %i octets\n", expect_send_size);
  printf("roundtrip count: %li\n", count);


  if ( (fd = socket(AF_UNIX, SOCK_STREAM, 0)) == -1) {
    perror("socket error");
    exit(-1);
  }

  memset(&addr, 0, sizeof(addr));
  addr.sun_family = AF_UNIX;
  if (*socket_path == '\0') {
    *addr.sun_path = '\0';
    strncpy(addr.sun_path+1, socket_path+1, sizeof(addr.sun_path)-2);
  } else {
    strncpy(addr.sun_path, socket_path, sizeof(addr.sun_path)-1);
  }

  printf ("debug: connet on socket_path: %s\n", socket_path);
  if (connect(fd, (struct sockaddr*)&addr, sizeof(addr)) == -1) {
    perror("connect error");
    exit(-1);
  }


 #ifdef HAS_CLOCK_GETTIME_MONOTONIC
    if (clock_gettime(CLOCK_MONOTONIC, &start) == -1) {
      perror("clock_gettime");
      return 1;
    }
#else
    if (gettimeofday(&start, NULL) == -1) {
      perror("gettimeofday");
      return 1;
    }
#endif

    int readcount=0;
    int writecount=0;

    for (i = 0; i < count; i++) {
      // printf("command argv[1]=%s\n", argv[1]);
      if ((writecount=write(fd, argv[1], expect_send_size)) != expect_send_size) {
        printf ("partial writecount %d\n" , writecount);
        perror("write");
        //return 1;
      }
      //printf ("write somthing\n");

      if ((readcount=read(fd, buf, expect_recv_size)) != expect_recv_size) {
        //printf ("partial readcount=%d \n" , readcount);
        printf ("partial readcount=%d: %.*s\n" , readcount, readcount, buf);

        perror("read");
        //return 1;
      }
      //printf ("read somthing\n");
    }

#ifdef HAS_CLOCK_GETTIME_MONOTONIC
    if (clock_gettime(CLOCK_MONOTONIC, &stop) == -1) {
      perror("clock_gettime");
      return 1;
    }

    delta = ((stop.tv_sec - start.tv_sec) * 1000000000 +
             (stop.tv_nsec - start.tv_nsec));

#else
    if (gettimeofday(&stop, NULL) == -1) {
      perror("gettimeofday");
      return 1;
    }

    delta =
        (stop.tv_sec - start.tv_sec) * 1000000000 + (stop.tv_usec - start.tv_usec) * 1000;

#endif

    printf("average latency: %li ns\n", delta / (count * 2));


  return 0;
}

File: sdn/ipc-bench-master/unix_lat_client.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

"""
和其他 _latency 程序不同, 需要和server一起使用
配合 ArsenalPython/stdmod/socket/pymotw 下


`python socket_echo_server_uds.py` 结果
配合 iproute2 tc `while true; do  make && ./tc/tc bench; done` 结果
"""
import socket
import sys
import time

usleep = lambda x: time.sleep(x / 1000000.0)

expect_recv_size = 0
expect_send_size = 0
BENCH_MODE = True
send_cmd = ""
bench_total = 0
# Create a UDS socket
sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)

# Connect the socket to the port where the server is listening
server_address = '/tmp/bench.socket'  # default socket

print ("debug: argc:%s argv:%s" % (len(sys.argv), sys.argv))
if len(sys.argv) != 5:
    print ("usage: ./unix_lat_client name.socket \"command\" <expect_recv_size> <roundtrip-count>");
    exit(1)
else:
    server_address = sys.argv[1]
    send_cmd = sys.argv[2]
    expect_send_size = len(send_cmd)
    expect_recv_size = int(sys.argv[3])
    bench_total = int(sys.argv[4])


if server_address.find("bench") == -1: # socket 字符参数中不包含 bench 以禁止bench模式
    print ("debug: not in BENCH_MODE, display all output ...\n")
    BENCH_MODE = True


print >> sys.stderr, 'connecting to %s' % server_address
try:
    sock.connect(server_address)
except socket.error, msg:
    print >> sys.stderr, msg
    sys.exit(1)

try:

    # 采样间隔时间, 默认0，不休眠
    polling_interval = 0
    # bench_total = 100000  # 测速 循环数 average latency: 3506 ns (4us) /
    polling_interval = 10
    #bench_total = 1000000

    bench_count = 0
    delta = 0
    total_bytes = 0
    start = time.time()

    while True:
        # Send data
        message = 'This is the message.  It will be repeated.'
        message = 'class show dev s1-eth2'  # len 323
        # print message


        if not BENCH_MODE:
            print >> sys.stderr, 'sending "%s"' % message

        # sock.sendall(message)
        sock.sendall(send_cmd)
        usleep(polling_interval)

        amount_received = 0
        amount_expected = 327  # len(message)

        # while amount_received < amount_expected:
        if True:
            data = sock.recv(expect_recv_size)
            amount_received += len(data)
            if BENCH_MODE:
                if ((bench_count % 100) == 0): print ".",
            else:
            #if True:
                print "recv len(data)=%s" % len(data)
                print >> sys.stderr, 'received "%s"' % data

        bench_count += 1
        total_bytes += len(data)
        if bench_count > bench_total:
            break

    if (BENCH_MODE) :
        print ""
        print "polling_interval: %s (等候数据回写间隔)" % polling_interval
        print "roundtrip count: %s" % bench_total
        print "total_bytes: %s" % total_bytes
        print "avg_bytes(should equals to interactive mode): %s bytes " % (total_bytes/bench_total)

    delta = time.time() - start
    print("average latency: %s ns" % int(delta / (bench_total * 2) * 1000000000))

finally:
    print >> sys.stderr, 'closing socket'
    sock.close()
